File: src/ltd/rust_lang/utils/queue/CircleQueue.java
package ltd.rust_lang.utils.queue;

import org.jetbrains.annotations.Contract;

public class CircleQueue<E> {
  private int size;
  private int front;
  private E[] elements;

  public CircleQueue() {
    elements = (E[]) new Object[10];
  }

  @Contract(pure = true)
  private int idx(int i) {
    var index = i + front;
    return index - (index >= elements.length ? elements.length : 0);
  }

  public int size() {
    return size;
  }

  public boolean isEmpty() {
    return size() == 0;
  }

  public void enQueue(E element) {
    ensure(size + 1);
    elements[idx(size)] = element;
    size++;
  }

  private void ensure(int capacity) {
    int old = elements.length;
    if (old >= capacity) return;
    var newC = old + (old >> 1);
    E[] newElements = (E[]) new Object[newC];
    for (int i = 0; i < size; ++i) {
      newElements[i] = elements[idx(i)];
    }
    elements = newElements;
    front = 0;
  }

  public E deQueue() {
    E frontEle = elements[front];
    elements[front] = null;
    front = idx(1);
    size--;
    return frontEle;
  }

  public E front() {
    return elements[front];
  }

  public void clear() {
    for (int i = 0; i < size; ++i) {
      elements[idx(i)] = null;
    }
    front = 0;
    size = 0;
  }

  @Override
  public String toString() {
    StringBuilder str = new StringBuilder();
    int len = elements.length;
    str.append("capacity = ")
        .append(len)
        .append(" size = ").append(size)
        .append(" front = ").append(front)
        .append(", [");
    for (int i = 0; i < len; ++i) {
      if (i != 0) str.append(", ");
      str.append(elements[i]);
    }
    str.append("]");
    return str.toString();
  }
}

File: src/main/kotlin/wiki/mdzz/utils/queue/CircleQueue.kt
package wiki.mdzz.utils.queue

class CircleQueue<E>(capacity: Int? = 10) {
    private var front: Int = 0
    private var size = 0
    private var elements: Array<Any?> = arrayOfNulls(capacity!!)

    fun enqueue(e: E?) {
        ensure(size + 1)
        elements[idx(size)] = e
        size++
    }

    fun dequeue(): E? {
        val frontE = elements[front]
        elements[front] = null
        front = idx(1)
        size--
        return frontE as E?
    }

    fun peek(): E? {
        return elements[front] as E?
    }

    fun clear() {
        for (i in 0 until size) {
            elements[idx(i)] = null
        }
        front = 0
        size = 0
    }

    fun size(): Int {
        return size
    }

    fun isEmpty(): Boolean {
        return size() == 0
    }

    private fun ensure(cap: Int) {
        val old = elements.size
        if (old >= cap) return
        val newC = old + (old shr 1)
        val newElements = arrayOfNulls<Any?>(newC)
        for (i in 0 until size) {
            newElements[i] = elements[idx(i)]
        }
        elements = newElements
        front = 0
    }

    override fun toString(): String {
        val str = StringBuilder()
        val len: Int = elements.size
        str.append("capacity = ")
            .append(len)
            .append(" size = ").append(size)
            .append(" front = ").append(front)
            .append(", [")
        for (i in 0 until len) {
            if (i != 0) str.append(", ")
            str.append(elements[i])
        }
        str.append("]")
        return str.toString()
    }

    private fun idx(i: Int): Int {
        return (front + i) % elements.size
    }
}
